viewer/viewer.js

```diff
@@ -7686,35 +7701,112 @@ function buildHuntOptions (huntId, hunt) {
+            return updateHuntStats(hunt, huntId, session, searchedSessions, cb);
+          });
+      });
+    }, function (err) { // done running a pass of the failed sessions
+      console.log('done with async failedSessionIds iteration', err);
+      console.log('changes?', changesSearchingFailedSessions);
+
+      function continueProcess () {
+        Db.setHunt(huntId, hunt, (err, info) => {
+          internals.runningHuntJob = undefined;
+          processHuntJobs(); // Start new hunt
+        });
+      }
+
+      if (hunt.failedSessionIds && hunt.failedSessionIds.length === 0) {
+        // we had failed sessions but we're done searching through them
+        // so we're completely done with this hunt (inital search and failed sessions)
+        hunt.status = 'finished';
+
+        if (hunt.notifier) {
+          let message = `*${hunt.name}* hunt job finished:\n*${hunt.matchedSessions}* matched sessions out of *${hunt.searchedSessions}* searched sessions`;
+          issueAlert(hunt.notifier, message, continueProcess);
+        } else {
+          return continueProcess();
+        }
+      } else if (hunt.failedSessionIds && hunt.failedSessionIds.length > 0 && changesSearchingFailedSessions) {
+        // there are still failed sessions, but there were also changes,
+        // so keep going
+        return continueProcess();
+      } else if (!changesSearchingFailedSessions) {
+        // there were no changes, we're still struggling to connect to one or
+        // more renote nodes, so error out
+        return pauseHuntJobWithError(huntId, hunt, { value: `Error hunting previously failed sessions: ${err}` });
+      }
+    });
+  }
+}
+
 // Actually do the search against ES and process the results.
 function runHuntJob (huntId, hunt, query, user) {
   let options = buildHuntOptions(huntId, hunt);
   let searchedSessions;
 
+  // look for failed sessions if we're done searching sessions normally
+  if (hunt.searchedSessions === hunt.totalSessions && hunt.failedSessionIds && hunt.failedSessionIds.length) {
+    huntFailedSessions(hunt, huntId, options, searchedSessions, user);
+  }
+
   Db.search('sessions2-*', 'session', query, { scroll: internals.esScrollTimeout }, function getMoreUntilDone (err, result) {
     if (err || result.error) {
       pauseHuntJobWithError(huntId, hunt, { value: `Hunt error searching sessions: ${err}` });
       return;
     }
 
     let hits = result.hits.hits;
-
     if (searchedSessions === undefined) {
       searchedSessions = hunt.searchedSessions || 0;
       // if the session query results length is not equal to the total sessions that the hunt
       // job is searching, update the hunt total sessions so that the percent works correctly
-      if (hunt.totalSessions !== (result.hits.total + searchedSessions)) {
-        hunt.totalSessions = result.hits.total + searchedSessions;
-      }
+      // TODO ECR - this borks rerunning through the failed sessions - the totalSessions gets incremented
```

Comment:
@awick What was the original purpose of this? Why was the percent not working correctly? Because of the scroll stuff?

If I keep this in when there are failed sessions, it just keeps incrementing the searched sessions.

Reply:
If I remember right, sessions could be added during the scroll so previous totalSessions would be wrong.

Reply:
👍 I left it in to see how it works with real data, since the test data never hits the scroll